edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/commands/KubernetesCrdCommand.cs

```diff
@@ -0,0 +1,263 @@
+                var secretMeta = new V1ObjectMeta(name: imagePullSecret.Key, namespaceProperty: this.deviceNamespace);
+                var newSecret = new V1Secret("v1", secretData, type: Constants.K8sPullSecretType, kind: "Secret", metadata: secretMeta);
+                Option<V1Secret> currentSecret;
+                try
+                {
+                    currentSecret = Option.Maybe(await this.client.ReadNamespacedSecretAsync(imagePullSecret.Key, this.deviceNamespace, cancellationToken: token));
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    Events.FailedToFindSecret(imagePullSecret.Key, ex);
+                    currentSecret = Option.None<V1Secret>();
+                }
+
+                try
+                {
+                    var v1Secret = await currentSecret.Match(
+                        async s =>
+                        {
+                            if ((s.Data != null) && s.Data.TryGetValue(Constants.K8sPullSecretData, out byte[] pullSecretData) &&
+                                pullSecretData.SequenceEqual(secretData[Constants.K8sPullSecretData]))
+                            {
+                                return s;
+                            }
+
+                            return await this.client.ReplaceNamespacedSecretAsync(
+                                newSecret,
+                                imagePullSecret.Key,
+                                this.deviceNamespace,
+                                cancellationToken: token);
+                        },
+                        async () => await this.client.CreateNamespacedSecretAsync(newSecret, this.deviceNamespace, cancellationToken: token));
+                    if (v1Secret == null)
+                    {
+                        throw new InvalidIdentityException("Image pull secret was not properly created");
+                    }
+                }
+                catch (Exception ex) when (!ex.IsFatal())
+                {
+                    Events.SecretCreateUpdateFailed(imagePullSecret.Key, ex);
+                }
+            }
+        }
+
+        public async Task ExecuteAsync(CancellationToken token)
+        {
+            string resourceName = this.iotHubHostname + Constants.K8sNameDivider + this.deviceId;
+            string metaApiVersion = Constants.K8sApi + "/" + Constants.K8sApiVersion;
+
+            var modulesList = new List<KubernetesModule<DockerConfig>>();
+            var secrets = new Dictionary<string, ImagePullSecret>();
+            foreach (var runtime in this.runtimeInfo)
+            {
+                foreach (var m in this.modules)
+                {
+                    var combinedConfig = this.combinedConfigProvider.GetCombinedConfig(m, runtime);
+                    CombinedDockerConfig dockerConfig = combinedConfig as CombinedDockerConfig;
```

Comment:
Is this cast necessary? Will GetCombinedConfig take a type parameter? If it is necessary, you should be checking the cast because IIRC, as will throw if it fails.

Reply:
I double checked, and the "as" operator returns null if conversion fails, and then I'm checking for null in the next line.  I think this could use some serious refactoring if we get different module types.